// @opd/g2plot-react/lib/.umi/umi.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("./core/polyfill");

require("@@/core/devScripts");

var _plugin = require("./core/plugin");

var _history = require("./core/history");

var _runtime = require("/Users/kagawa/workspace/open-data-plan/g2plot-react/node_modules/@umijs/runtime");

var _index = require("/Users/kagawa/workspace/open-data-plan/g2plot-react/node_modules/@umijs/renderer-react/dist/index.js");

var getClientRender = function getClientRender() {
  var args = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return _plugin.plugin.applyPlugins({
    key: 'render',
    type: _runtime.ApplyPluginsType.compose,
    initialValue: function initialValue() {
      return (0, _index.renderClient)({
        // @ts-ignore
        routes: require('./core/routes').routes,
        plugin: _plugin.plugin,
        history: (0, _history.createHistory)(args.hot),
        rootElement: 'root',
        defaultTitle: 'G2Plot React'
      });
    },
    args: args
  });
};

var clientRender = getClientRender();

var _default = clientRender();

exports.default = _default;
window.g_umi = {
  version: '3.1.1'
}; // hot module replacement
// @ts-ignore

if (module.hot) {
  // @ts-ignore
  module.hot.accept('./core/routes', function () {
    getClientRender({
      hot: true
    })();
  });
}
